### DISCUSSION TITLE
Homebrew installation fails because it tries to install wrong version of cask

### DISCUSSION THREAD

**Discussion by fasmat** [Homebrew/discussions#1672] [Everyday usage]
Hi everyone,

I have my own tap of a few casks one of which fails to install for the last 2 days. It seems as if homebrew tries to install the wrong version (978.2j) instead of the version specified in the cask (981.3b).

Cask code here: https://github.com/fasmat/homebrew-trader-workstation/blob/main/Casks/trader-workstation.rb

Output of `brew install trader-workstation`:

<details><summary>Click to expand</summary><pre><code>

==> Cask::Installer#install
==> Printing caveats
==> Cask::Installer#fetch
==> Downloading https://download2.interactivebrokers.com/installers/tws/stable-standalone/tws-stable-standalone-macosx-x64.dmg
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/3.1.11-45-g33e71c9\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.7\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --silent --head --request GET https://download2.interactivebrokers.com/installers/tws/stable-standalone/tws-stable-standalone-macosx-x64.dmg
Already downloaded: /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg
==> Checking quarantine support
/usr/bin/xattr
/usr/bin/swift /usr/local/Homebrew/Library/Homebrew/cask/utils/quarantine.swift
==> Quarantine is available.
==> Verifying Gatekeeper status of /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg
/usr/bin/xattr -p com.apple.quarantine /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg
==> /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg is quarantined
Warning: No checksum defined for cask 'trader-workstation', skipping verification.
hdiutil imageinfo -format /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg
==> Installing Cask trader-workstation
==> Cask::Installer#stage
==> Extracting primary container
==> Using container class UnpackStrategy::Dmg for /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg
hdiutil attach -plist -nobrowse -readonly -mountrandom /private/tmp/d20210611-65073-12xr6ww /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg
find . -print0
mkbom -s -i /private/tmp/20210611-65073-aq0cxz.list -- /private/tmp/20210611-65073-a4k380.bom
ditto --bom /private/tmp/20210611-65073-a4k380.bom -- /private/tmp/d20210611-65073-12xr6ww/dmg.6T1kl7 /private/tmp/d20210611-65073-1lef23u
diskutil info -plist /private/tmp/d20210611-65073-12xr6ww/dmg.6T1kl7
diskutil eject /private/tmp/d20210611-65073-12xr6ww/dmg.6T1kl7
cp -pR /private/tmp/d20210611-65073-1lef23u/Trader\ Workstation\ 981\ Installer.app/. /usr/local/Caskroom/trader-workstation/978.2j/Trader\ Workstation\ 981\ Installer.app
chmod -Rf +w /private/tmp/d20210611-65073-1lef23u
==> Verifying Gatekeeper status of /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg
/usr/bin/xattr -p com.apple.quarantine /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg
==> /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg is quarantined
==> Propagating quarantine from /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg to /usr/local/Caskroom/trader-workstation/978.2j
/usr/bin/xattr -p com.apple.quarantine /Users/mafa/Library/Caches/Homebrew/downloads/ecef478e8a2f6f113371ee5fa27304a1c097e5df8f8c33bbca343713f9b9f8eb--tws-stable-standalone-macosx-x64.dmg
/usr/bin/xargs -0 -- /bin/chmod -h u\+w
/usr/bin/xargs -0 -- /usr/bin/xattr -w com.apple.quarantine 0181\;60c325c4\;Homebrew\\x20Cask\;D2F6D198-0541-4D45-8B29-4B3E4E39DEB5
==> Creating metadata directory: /usr/local/Caskroom/trader-workstation/.metadata/978.2j/20210611091018.093
==> Creating metadata subdirectory: /usr/local/Caskroom/trader-workstation/.metadata/978.2j/20210611091018.093/Casks
==> Installing artifacts
==> 2 artifacts defined
#<SortedSet: {#<Cask::Artifact::Uninstall:0x00007fed3c24a7a8 @cask=#<Cask::Cask:0x00007fed3c0dfc38 @token="trader-workstation", @sourcefile_path=#<Pathname:/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Casks/trader-workstation.rb>, @tap=#<Tap:0x00007fed580f03f0 @user="Homebrew", @repo="cask", @name="homebrew/cask", @full_name="Homebrew/homebrew-cask", @path=#<Pathname:/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask>, @alias_table=nil, @alias_reverse_table=nil, @formula_renames={}, @cask_dir=#<Pathname:/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Casks>>, @block=#<Proc:0x00007fed3c0df940@/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Casks/trader-workstation.rb:1>, @default_config=#<Cask::Config:0x00007fed5817bb08 @explicit={}>, @caskroom_path=#<Pathname:/usr/local/Caskroom/trader-workstation>, @metadata_master_container_path=#<Pathname:/usr/local/Caskroom/trader-workstation/.metadata>, @config=#<Cask::Config:0x00007fed3c240d98 @explicit={}>, @dsl=#<Cask::DSL:0x00007fed3c240af0 @cask=#<Cask::Cask:0x00007fed3c0dfc38 ...>, @token="trader-workstation", @version="978.2j", @sha256=:no_check, @url=#<URL::DSL:0x00007fed3c240488 @uri=#<URI::HTTPS https://download2.interactivebrokers.com/installers/tws/stable-standalone/tws-stable-standalone-macosx-x64.dmg>, @verified=nil, @using=nil, @tag=nil, @branch=nil, @revisions=nil, @revision=nil, @trust_cert=nil, @cookies=nil, @referer=nil, @header=nil, @user_agent=:default, @data=nil, @specs={:user_agent=>:default}>, @name=["Trader Workstation"], @desc="Trading software", @homepage="https://www.interactivebrokers.com/", @staged_path=#<Pathname:/usr/local/Caskroom/trader-workstation/978.2j>, @artifacts=#<SortedSet: {...}>, @language_eval=nil>>, @directives={:quit=>"com.install4j.5889-6375-8446-2021.22", :script=>{:executable=>"/Users/mafa/Applications/Trader Workstation 978/Trader Workstation 978 Uninstaller.app/Contents/MacOS/JavaApplicationStub", :args=>["-q"]}, :signal=>[]}>, #<Cask::Artifact::Installer:0x00007fed3c24b2c0 @cask=#<Cask::Cask:0x00007fed3c0dfc38 @token="trader-workstation", @sourcefile_path=#<Pathname:/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Casks/trader-workstation.rb>, @tap=#<Tap:0x00007fed580f03f0 @user="Homebrew", @repo="cask", @name="homebrew/cask", @full_name="Homebrew/homebrew-cask", @path=#<Pathname:/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask>, @alias_table=nil, @alias_reverse_table=nil, @formula_renames={}, @cask_dir=#<Pathname:/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Casks>>, @block=#<Proc:0x00007fed3c0df940@/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Casks/trader-workstation.rb:1>, @default_config=#<Cask::Config:0x00007fed5817bb08 @explicit={}>, @caskroom_path=#<Pathname:/usr/local/Caskroom/trader-workstation>, @metadata_master_container_path=#<Pathname:/usr/local/Caskroom/trader-workstation/.metadata>, @config=#<Cask::Config:0x00007fed3c240d98 @explicit={}>, @dsl=#<Cask::DSL:0x00007fed3c240af0 @cask=#<Cask::Cask:0x00007fed3c0dfc38 ...>, @token="trader-workstation", @version="978.2j", @sha256=:no_check, @url=#<URL::DSL:0x00007fed3c240488 @uri=#<URI::HTTPS https://download2.interactivebrokers.com/installers/tws/stable-standalone/tws-stable-standalone-macosx-x64.dmg>, @verified=nil, @using=nil, @tag=nil, @branch=nil, @revisions=nil, @revision=nil, @trust_cert=nil, @cookies=nil, @referer=nil, @header=nil, @user_agent=:default, @data=nil, @specs={:user_agent=>:default}>, @name=["Trader Workstation"], @desc="Trading software", @homepage="https://www.interactivebrokers.com/", @staged_path=#<Pathname:/usr/local/Caskroom/trader-workstation/978.2j>, @artifacts=#<SortedSet: {...}>, @language_eval=nil>>, @args={:must_succeed=>true, :sudo=>false, :args=>["-q"], :print_stdout=>true}, @path=#<Pathname:/usr/local/Caskroom/trader-workstation/978.2j/Trader Workstation 978 Installer.app/Contents/MacOS/JavaApplicationStub>>}>
==> Installing artifact of class Cask::Artifact::Installer
==> Running installer script '/usr/local/Caskroom/trader-workstation/978.2j/Trader Workstation 978 Installer.app/Contents/MacOS/JavaApplicationStub'
/usr/bin/env PATH=/usr/local/bin:/usr/local/sbin:/usr/local/Homebrew/Library/Homebrew/shims/scm:/usr/bin:/bin:/usr/sbin:/sbin /usr/local/Caskroom/trader-workstation/978.2j/Trader\ Workstation\ 978\ Installer.app/Contents/MacOS/JavaApplicationStub -q
env: /usr/local/Caskroom/trader-workstation/978.2j/Trader Workstation 978 Installer.app/Contents/MacOS/JavaApplicationStub: No such file or directory
==> Purging files for version 978.2j of Cask trader-workstation
Error: Failure while executing; `/usr/bin/env PATH=/usr/local/bin:/usr/local/sbin:/usr/local/Homebrew/Library/Homebrew/shims/scm:/usr/bin:/bin:/usr/sbin:/sbin /usr/local/Caskroom/trader-workstation/978.2j/Trader\ Workstation\ 978\ Installer.app/Contents/MacOS/JavaApplicationStub -q` exited with 127. Here's the output:
env: /usr/local/Caskroom/trader-workstation/978.2j/Trader Workstation 978 Installer.app/Contents/MacOS/JavaApplicationStub: No such file or directory

/usr/local/Homebrew/Library/Homebrew/system_command.rb:302:in 'assert_success!'
/usr/local/Homebrew/Library/Homebrew/system_command.rb:65:in 'run!'
/usr/local/Homebrew/Library/Homebrew/system_command.rb:40:in 'run'
/usr/local/Homebrew/Library/Homebrew/system_command.rb:44:in 'run!'
/usr/local/Homebrew/Library/Homebrew/cask/artifact/installer.rb:38:in 'install_phase'
/usr/local/Homebrew/Library/Homebrew/cask/installer.rb:233:in 'block in install_artifacts'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/lib/ruby/2.6.0/set.rb:777:in 'each'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/lib/ruby/2.6.0/set.rb:777:in 'each'
/usr/local/Homebrew/Library/Homebrew/cask/installer.rb:226:in 'install_artifacts'
/usr/local/Homebrew/Library/Homebrew/cask/installer.rb:111:in 'install'
/usr/local/Homebrew/Library/Homebrew/cask/cmd/install.rb:69:in 'block in install_casks'
/usr/local/Homebrew/Library/Homebrew/cask/cmd/install.rb:68:in 'each'
/usr/local/Homebrew/Library/Homebrew/cask/cmd/install.rb:68:in 'install_casks'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:167:in 'install'
/usr/local/Homebrew/Library/Homebrew/brew.rb:122:in '&lt;main&gt;'

</code></pre></details>

Any ideas what the issue could be?

Thanks in advance,
fasmat

**Comment by gromgit** [comment]
What's the output of `brew info trader-workstation`?

**Reply by fasmat** [reply]
Ah! Thanks for the input, apparently it tries to install a different version of trader-workstation:

<details><summary>Click to expand</summary><pre><code>

trader-workstation: 978.2j
https://www.interactivebrokers.com/
Not installed
From: https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/trader-workstation.rb
==> Name
Trader Workstation
==> Description
Trading software
==> Artifacts
/usr/local/Caskroom/trader-workstation/978.2j/Trader Workstation 978 Installer.app/Contents/MacOS/JavaApplicationStub (Installer)
==> Analytics
install: 7 (30 days), 34 (90 days), 133 (365 days)

</code></pre></details>

I'll check out the official homebrew-cask repository, seems there is an outdated version of the software
